### PR TITLE
grpc service

### PR DESCRIPTION
**Pull request by majst01**
In the long run we want to replace the ipam library usage in metal-api with a dedicated microservice. With this PR we provide a grpc service for this purpose.

replaces #46

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -31,3 +36,28 @@ jobs:
...
+    - name: Extract metadata (tags, labels) for Docker
+      id: meta
+      uses: docker/metadata-action@v3
+      with:
+        images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+        tags: |
+          type=ref,event=branch
+          type=ref,event=pr
+          type=semver,pattern={{version}}
+          type=semver,pattern={{major}}.{{minor}}
```
Comment (Gerrit91) [inline .github/workflows/main.yml:56]: Does not produce any latest tags anymore and also changes general style of our PR image names. Just stealing from metal-api or something alike makes it shorter, keeps existing tags and only requires a single workflow file.

File: cmd/client/client.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+						Action: func(ctx *cli.Context) error {
+							c := client(ctx)
+							result, err := c.ReleaseChildPrefix(context.Background(), connect.NewRequest(&v1.ReleaseChildPrefixRequest{
+								Cidr: ctx.String("cidr"),
+							}))
+
+							if err != nil {
+								return err
+							}
+							fmt.Printf("child prefix:%q from %q released\n", result.Msg.Prefix.Cidr, result.Msg.Prefix.ParentCidr)
```
Comment (Gerrit91) [inline cmd/client/client.go:100]: Don't know how likely this is, but potentially there can be nil pointer panics...

File: pkg/service/ipam-service.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+func New(log *zap.SugaredLogger, ipamer goipam.Ipamer) *IPAMService {
+	return &IPAMService{
+		log:    log,
+		ipamer: ipamer,
+	}
+}
+
+func (i *IPAMService) CreatePrefix(_ context.Context, req *connect.Request[v1.CreatePrefixRequest]) (*connect.Response[v1.CreatePrefixResponse], error) {
+	i.log.Debugw("createprefix", "req", req)
+	// FIXME context must be passed here
```
Comment (Gerrit91) [inline pkg/service/ipam-service.go:28]: Will this become a follow-up PR?

File: pkg/service/ipam-service.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	resp, err := i.ipamer.NewPrefix(req.Msg.Cidr)
```
Comment (Gerrit91) [inline pkg/service/ipam-service.go:29]: Are there any plans on request validation? Would be nice to do it here before separate DB layers start to implement this individually.

File: pkg/service/ipam-service.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	if err != nil {
+		i.log.Errorw("listprefixes", "error", err)
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	var result []*v1.Prefix
+	for _, cidr := range resp {
+		p := i.ipamer.PrefixFrom(cidr)
+		if p == nil {
+			continue
```
Comment (Gerrit91) [inline pkg/service/ipam-service.go:90]: Is this happening usually? Maybe worth logging a warning?

**Review by Gerrit91 — APPROVED** [review]
Looks fine to me, only minor comments added.

File: pkg/service/ipam-service.go (hunk 4 above)
Comment (majst01) [reply pkg/service/ipam-service.go:29]: It is already validated in the impl which is in front of the db layers, see prefix.go

File: pkg/service/ipam-service.go (hunk 3 above)
Comment (majst01) [reply pkg/service/ipam-service.go]: FIXME removed, opened #83

File: pkg/service/ipam-service.go (hunk 4 above)
Comment (majst01) [reply pkg/service/ipam-service.go:28]: done

File: cmd/client/client.go (hunk 2 above)
Comment (majst01) [reply cmd/client/client.go:103]: done

File: .github/workflows/main.yml (hunk 1 above)
Comment (majst01) [reply .github/workflows/main.yml]: also copied over from api-server

**Pull request merged**